Access table associations using the dot syntax

sdustinh opened the pull request:
Reference: #11366 #11369

sdustinh committed:
Added support for dot syntax on Table::association()

Reference: #11366

sdustinh committed:
Added test for association dot syntax format

Reference: #11366

ADmad commented:
Instead of modifying `associations()` wouldn't it be better to add a new `getAssociations()` method which supports the new dot syntax and deprecate the former? Two birds with one stone.

sdustinh commented:
:+1: That would be awesome. I'll add it.

ADmad commented on src/ORM/Table.php:
```diff
@@ -872,14 +872,27 @@ public function hasBehavior($name)
     public function association($name)
     {
-        return $this->_associations->get($name);
+        list($name, $next) = array_pad(explode('.', $name, 2), 2, null);
```
Optimization should be added for checking whether name contains `.` and return early from the method.

dereuromark commented:
But singular association() => getAssociation() afaik.

ADmad commented:
> But singular association() => getAssociation() afaik.

Yeah, sorry I typoed in previous comment.

markstory commented on tests/TestCase/ORM/TableTest.php:
```diff
@@ -511,6 +511,30 @@ function (Event $event, $query, $options) use ($expected) {
+        $members->belongsToMany('Groups');
+
+        $association = $groups->association('GroupsMembers.Members.Groups');
+        $this->assertInstanceOf('Cake\ORM\Association\BelongsToMany', $association);
```
You could use `BelongsToMany::class` here 🏌️

sdustinh committed:
Deprecated Table::association()

sdustinh commented:
@ADmad This should be good to go unless you guys see anything you want changed.

ADmad commented on src/ORM/Table.php:
```diff
@@ -874,11 +874,40 @@ public function hasBehavior($name)
+     */
+    public function getAssociation($name)
+    {
+        if (strpos($name, '.') !== false) {
```
We normally follow "return early" paradigm so this should instead check for false and return the association.

sdustinh replied:
No problem. I'll take care of it now.

sdustinh committed:
Return early if dot syntax is not being used.

ADmad commented on src/ORM/Table.php:
```diff
@@ -874,12 +874,41 @@ public function hasBehavior($name)
+        $result = $this->_associations->get($name);
+
+        if ($result !== null && $next !== null) {
+            $result = $result->getTarget()->association($next);
```
You are using deprecated `association()` method :smile:

sdustinh replied:
Haha, jesus.

sdustinh committed:
Remove call to deprecated association() method

markstory commented on src/ORM/Table.php:
```diff
@@ -874,12 +874,41 @@ public function hasBehavior($name)
     public function association($name)
     {
-        return $this->_associations->get($name);
+        return $this->getAssociation($name);
```
Could we get a `deprecationWarning()` here? It'll save some time in the future.

sdustinh committed:
Trigger a deprecation warning

sdustinh commented:
@markstory Added a deprecation warning. I wasn't sure if you guy were adding tests for the deprecation warning or not. If you want that, I would be happy to add that as well.

markstory commented:
>  I wasn't sure if you guy were adding tests for the deprecation warning or not. If you want that, I would be happy to add that as well.

We do have tests for deprecated methods, but we're also in the process of updating all the internal usage of deprecated methods so that they are unused.

sdustinh commented:
@markstory Would you like me to replace all instances of `Table::association()` with `Table::getAssociation()` in this PR as well? If so, should this include those used in the tests?

markstory commented:
@sdustinh That would be ideal. All uses in tests should be updated except one that is a test for the association() method to help prevent regressions until 4.0

sdustinh commented:
@markstory I replaced all instances of `Table::association()` with `Table::getAssociation()` in both `src/` and `tests/`. I wasn't able to find a test specific for `Table::association()`. For now, I'll just push up those changes for review and if there is a test in there that I missed I'll patch that after.

sdustinh committed:
Updated instance of Table::association()

Replaced all instances of Table::association() with Table::getAssociation() in the framework source and tests.

markstory commented:
Looking good :+1:
